Feat/benchmark/graphics (#1507)
edgelesssys/constellation: docs/docs/overview/performance.md

Similarly, AMD and Google jointly released a [performance benchmark](https://www.amd.com/system/files/documents/3rd-gen-epyc-gcp-c2d-conf-compute-perf-brief.pdf) for CVMs based on 3rd Gen AMD EPYC processors (Milan) with SEV-SNP. With high-performance computing workloads like WRF, NAMD, Ansys CFS, and Ansys LS_DYNA, they found similar results with only small (2%--4%) performance degradation compared to standard VMs. You can expect to see similar performance for compute-intensive workloads running with Constellation GCP.

## Performance impact from I/O and network

To assess the overall performance of Constellation, we benchmarked Constellation v2.6.0 in terms of storage I/O using [`fio`](https://fio.readthedocs.io/en/latest/fio_doc.html) and network performance using the [Kubernetes Network Benchmark](https://github.com/InfraBuilder/k8s-bench-suite#knb--kubernetes-network-be)

> We tested Constellation on Azure and GCP and compared the results against the respective managed Kubernetes offerings, AKS, and GKE.

Author: I also removed "respective", then we can omit the first comma.

The tests use [`iperf`](https://iperf.fr/) to measure the bandwidth.

Constellation on GCP and GKE had a maximum network bandwidth of [10 Gbps](https://cloud.google.com/compute/docs/general-purpose-machines#n2d_machineshttps://cloud.google.com/compute/docs/general-purpose-machines#n2d_machines).
Constellation on Azure had a maximum network bandwidth of [1.25 Gbps](https://learn.microsoft.com/en-us/azure/virtual-machines/dcasv5-dcadsv5-series#dcasv5-series-products)
AKS had a maximum network bandwidth of [12.5 Gbps](https://learn.microsoft.com/en-us/azure/virtual-machines/dasv5-dadsv5-series#dasv5-series).
Constellation on Azure and AKS used a 1500 MTU.
Constellation on GCP used an 8896 MTU. GKE used a 1450 MTU.

The difference can largely be attributed to two factors.

1. Constellation's [network encryption](../architecture/networking.md) via Cilium and WireGuard that protects data in-transit.
2. [AMD SEV using SWIOTLB bounce buffers](https://lore.kernel.org/all/20200204193500.GA15564@ashkalra_ubuntu_server/T/) for all DMA including network I/O.

##### Pod-to-Pod

In this scenario, the client Pod connects directly to the server pod in its IP address.

> ![Pod2Pod concept](../_media/benchmark_p2p_concept.webp)

Author: Unfortunately not. I should redo them myself

The results for "Pod-to-Pod" TCP are as follows:

![Network Pod2Pod TCP benchmark graph](../_media/benchmark_net_p2p_tcp.png)

The results for "Pod-to-Pod" UDP are as follows:

![Network Pod2Pod UDP benchmark graph](../_media/benchmark_net_p2p_udp.png)

##### Pod-to-Service

Tn this scenario, the client Pod connects to the server Pod via a ClusterIP service. This is more relevant to real-world use cases.

![Pod2Pod concept](../_media/benchmark_p2svc_concept.webp)

The results for “Pod-to-Service” TCP are as follows:

![Network Pod2SVC TCP benchmark graph](../_media/benchmark_net_p2svc_tcp.png)

The results for “Pod-to-Service” UDP are as follows:

![Network Pod2SVC TCP benchmark graph](../_media/benchmark_net_p2svc_udp.png)

> Comparing Constellation on GCP with GKE, Constellation has 58% less TCP bandwidth less but a similar UDP bandwidth.

Reviewer: ```suggestion
Comparing Constellation on GCP with GKE, Constellation has 58% less TCP bandwidth but a similar UDP bandwidth.
```

> Constellation on Azure isn't directly comparable as the Confidential VM types in Azure have only 10% of the regular VM types' bandwidth.

Reviewer: ```suggestion
Constellation on Azure isn't directly comparable as Azure's Confidential VMs have only 10% of the regular VMs' bandwidth.
```
nit: maybe this is a bit simpler

> Constellation's network encryption adds almost no overhead on top of that limitation.

Reviewer: Isn't the overhead around 30%?

Author: It should be accurate now. ~58% loss on GCP, ~15% loss on Azure

#### Storage I/O

Azure and GCP offer persistent storage for their Kubernetes services AKS and GKE via the Container Storage Interface (CSI). CSI storage in Kubernetes is available via `PersistentVolumes` (PV) and consumed via `PersistentVolumeClaims` (PVC).

> Upon requesting persistent storage through a PVC, GKE, and AKS will provision a PV as defined by a default [storage class](https://kubernetes.io/docs/concepts/storage/storage-classes/).

Reviewer: ```suggestion
Upon requesting persistent storage through a PVC, GKE and AKS will provision a PV as defined by a default [storage class](https://kubernetes.io/docs/concepts/storage/storage-classes/).
```
FP

Constellation provides persistent storage on Azure and GCP [that's encrypted on the CSI layer](../architecture/encrypted-storage.md).
Similarly, upon a PVC request, Constellation will provision a PV via a default storage class.
Specifically, on Azure the [Azure Disk CSI driver with encryption](https://github.com/edgelesssys/constellation-azuredisk-csi-driver) v1.1.2 was used.

> On GCP the [Compute Engine Persistent Disk CSI Driver with encryption](https://github.com/edgelesssys/constellation-gcp-compute-persistent-disk-csi-driver) v1.1.2.

Reviewer: These sentences seem a bit misplaced or lack context.

Author: Removed that entirely. It's fixed with the Constellation version anyway. So this is probably just confusing.

> For Constellation on Azure and AKS, we ran the benchmark with Azure Disk storage [Standard SSD](https://learn.microsoft.com/en-us/azure/virtual-machines/disks-types#standard-ssds) of 400 GB size.

Reviewer: We have mixed GB/GiB, MB/MiB etc on this page. Maybe you can check or unify. But don't put to much effort in finding out. It's not that important.

Author: Checked and updated. We use GiB and MB/s, which is what the docs of Azure and Google do and what we are using in the benchmark configurations.

The [DC4as machine type](https://learn.microsoft.com/en-us/azure/virtual-machines/dasv5-dadsv5-series#dasv5-series) with four cores provides the following maximum performance:
- 6400 (20000 burst) IOPS

> - 144 MB/s (600 MB/s burst) throughput
> However, the performance is bound by the capabilities of the 512 GiB Standard SSD size:

Reviewer: ```suggestion
- 144 MB/s (600 MB/s burst) throughput

However, the performance is bound by the capabilities of the 512 GiB Standard SSD size:
```

- 500 (600 burst) IOPS
- 60 MB/s (150 MB/s burst) throughput

For Constellation on GCP and GKE, we ran the benchmark with Compute Engine Persistent Disk Storage [pd-standard](https://cloud.google.com/compute/docs/disks) of 400 GB size.

> Our N2D machine type with four cores and pd-standard provides the following [maximum performance](https://cloud.google.com/compute/docs/disks/performance#n2d_vms):

Reviewer: ```suggestion
The N2D machine type with four cores and pd-standard provides the following [maximum performance](https://cloud.google.com/compute/docs/disks/performance#n2d_vms):
```

- 15,000 write IOPS
- 3,000 read IOPS
- 240 MB/s write throughput
- 240 MB/s read throughput

The [`fio`](https://fio.readthedocs.io/en/latest/fio_doc.html) benchmark consists of several tests.

> We utilized [`Kubestr`](https://github.com/kastenhq/kubestr) to run `fio` in Kubernetes.

Reviewer: ```suggestion
We used [`Kubestr`](https://github.com/kastenhq/kubestr) to run `fio` in Kubernetes.
```
nit https://learn.microsoft.com/en-us/style-guide/word-choice/use-simple-words-concise-sentences

The default test performs randomized access patterns that accurately depict worst-case I/O scenarios for most applications.

The following `fio` settings were used:

- No Cloud caching
- No OS caching
- Single CPU
- 60 seconds runtime
- 10 seconds ramp-up time
- 10 GB file
- IOPS: 4 KB blocks and 128 iodepth
- Bandwidth: 1024 KB blocks and 128 iodepth

> You can find the `fio` test configuration [here](../../../.github/actions/e2e_benchmark/fio.ini).

Reviewer: ```suggestion
For more details, see the [`fio` test configuration](../../../.github/actions/e2e_benchmark/fio.ini).
```
or some other phrasing that gives a better link text

The results for "Rand Read" IOPS are as follows:

![I/O read IOPS benchmark graph](../_media/benchmark_fio_read_iops.png)

The results for "Rand Write" IOPS are as follows:

![I/O write IOPS benchmark graph](../_media/benchmark_fio_write_iops.png)

The results for "Rand Read" bandwidth are as follows:

![I/O read bandwidth benchmark graph](../_media/benchmark_fio_read_bw.png)

The results for "Rand Write" bandwidth are as follows:

![I/O write bandwidth benchmark graph](../_media/benchmark_fio_write_bw.png)

> Comparing Constellation on GCP with GKE, you see that Constellation has a similar read but 75% less write bandwidth.

Reviewer: ```suggestion
Comparing Constellation on GCP with GKE, Constellation has a similar read but 75% less write bandwidth.
```

> Constellation on Azure has 15% less read and 54% less write bandwidth compared to AKS.

Reviewer: I can't reproduce these numbers from the graphs.

Author: Better now?

AMD and Azure jointly released a [performance benchmark](https://community.amd.com/t5/business/microsoft-azure-confidential-computing-powered-by-3rd-gen-epyc/ba-p/497796) for CVMs based on 3rd Gen AMD EPYC processors (Milan) with SEV-SNP. With a range of mostly compute-intensive benchmarks like SPEC CPU 2017 and CoreMark, they found that CVMs only have a small (2%--8%) performance degradation compared to standard VMs. You can expect to see similar performance for compute-intensive workloads running with Constellation on Azure.

Similarly, AMD and Google jointly released a [performance benchmark](https://www.amd.com/system/files/documents/3rd-gen-epyc-gcp-c2d-conf-compute-perf-brief.pdf) for CVMs based on 3rd Gen AMD EPYC processors (Milan) with SEV-SNP. With high-performance computing workloads like WRF, NAMD, Ansys CFS, and Ansys LS_DYNA, they found similar results with only small (2%--4%) performance degradation compared to standard VMs. You can expect to see similar performance for compute-intensive workloads running with Constellation on GCP.

> ## Performance impact from I/O and network

Reviewer: ```suggestion
## Performance impact from storage and network
```

We used the following infrastructure configurations.

- Nodes: 2 (2 Worker)
- Machines: `n2d-standard-4` 2nd Generation AMD EPYC (Rome) processor with 4 Cores, 16 GiB of memory
- CVM: `false`
- Zone: `europe-west3-b`

### Results

#### Network

We performed a thorough analysis of the network performance of Constellation, specifically focusing on measuring TCP and UDP bandwidth.
The benchmark measured the bandwidth of pod-to-pod and pod-to-service connections between two different nodes using [`iperf`](https://iperf.fr/).

GKE and Constellation on GCP had a maximum network bandwidth of [10 Gbps](https://cloud.google.com/compute/docs/general-purpose-machines#n2d_machineshttps://cloud.google.com/compute/docs/general-purpose-machines#n2d_machines).
Constellation on Azure had a maximum network bandwidth of [1.25 Gbps](https://learn.microsoft.com/en-us/azure/virtual-machines/dcasv5-dcadsv5-series#dcasv5-series-products).

> AKS had a maximum network bandwidth of [12.5 Gbps](https://learn.microsoft.com/en-us/azure/virtual-machines/dasv5-dadsv5-series#dasv5-series).

Reviewer: Should we make the meaning and consequence of this clearer already here? So people are less negatively surprised if they see the graph?
But I'm not sure if this helps. Many people probably skim the page and just notice the graphs. Can we add a hint inside the graphs that this hits a limitation of the CVMs? Or can we add AKS with CVMs? (Does it have the same limitation?) Or both add AKS with CVMs and the limitation hint?

Author: I've discussed with Felix and grouped the graphs by cloud provider for now while running AKS with CVMs.

Constellation on Azure and AKS used an MTU of 1500.
Constellation on GCP used an MTU of 8896. GKE used an MTU of 1450.

The difference in network bandwidth can largely be attributed to two factors.

* Constellation's [network encryption](../architecture/networking.md) via Cilium and WireGuard, which protects data in-transit.
* [AMD SEV using SWIOTLB bounce buffers](https://lore.kernel.org/all/20200204193500.GA15564@ashkalra_ubuntu_server/T/) for all DMA including network I/O.

##### Pod-to-Pod

In this scenario, the client Pod connects directly to the server pod via its IP address.

```mermaid
flowchart LR
    subgraph Node A
    Client[Client]
    end
    subgraph Node B
    Server[Server]
    end
    Client ==>|traffic| Server
```

The results for "Pod-to-Pod" TCP are as follows:

![Network Pod2Pod TCP benchmark graph](../_media/benchmark_net_p2p_tcp.png)

The results for "Pod-to-Pod" UDP are as follows:

![Network Pod2Pod UDP benchmark graph](../_media/benchmark_net_p2p_udp.png)

##### Pod-to-Service

> Tn this scenario, the client Pod connects to the server Pod via a ClusterIP service. This is more relevant to real-world use cases.

Reviewer: ```suggestion
In this scenario, the client Pod connects to the server Pod via a ClusterIP service. This is more relevant to real-world use cases.
```

The results for "Rand Read" IOPS are as follows:

![I/O read IOPS benchmark graph](../_media/benchmark_fio_read_iops.png)

The results for "Rand Write" IOPS are as follows:

![I/O write IOPS benchmark graph](../_media/benchmark_fio_write_iops.png)

The results for "Rand Read" bandwidth are as follows:

![I/O read bandwidth benchmark graph](../_media/benchmark_fio_read_bw.png)

The results for "Rand Write" bandwidth are as follows:

![I/O write bandwidth benchmark graph](../_media/benchmark_fio_write_bw.png)

On GCP, we can see that we exceed the maximum performance guarantees of the chosen disk type.
There are two possible explanations for this. (1) There is some cloud caching in place we don't control. (2) The underlying provisioned disk size is larger than the requested on, which would yield higher performance boundaries.

Comparing Constellation on GCP with GKE, Constellation has a similar bandwidth but ~10% less IOPS performance.
Constellation on Azure has a similar IOPS performance compared to AKS, where both probably hit the maximum storage performance. Constellation has ~15% less read and write bandwidth.

## Conclusion

> Despite providing substantial [security benefits](./security-benefits.md), Constellation overall only has a slight performance overhead over the managed Kubernetes offerings AKS and GKE. Constellation is on par in most benchmarks but is slightly slower in certain scenarios due to network and storage encryption.

Author: Updated the conclusions
